### PR TITLE
Code for 5 in development

### PR DESCRIPTION
**Pull request by ablqk**
Code written alongside the chapter.

- First commit contains the first part (5.1 business definitions)
- Second commit contains the second part (5.2 whatever is it)

TODO on this second part: I'm missing 2 tests. Let's pair on them at some point, because the current code in main is too far from what I expect to write there.

### CODE REVIEW

File: chapter-05/1_1_types/money/convert_test.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	}{
+		"34.98 USD to EUR": {
+			amount: money.Amount{},
+			to:     money.Currency{},
+			validate: func(t *testing.T, got money.Amount, err error) {
+				if err != nil {
+					t.Errorf("expected no error, got %s", err.Error())
+				}
+				expected := money.Amount{}
+				if !reflect.DeepEqual(got, expected) {
```
Comment (floppyzedolfin) [inline chapter-05/1_1_types/money/convert_test.go]: why do we need deepequal here?

File: chapter-05/1_1_types/money/currency.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package money
+
+// Currency defines the code of a currency.
+type Currency struct {
+	code string
```
Comment (floppyzedolfin) [inline chapter-05/1_1_types/money/currency.go]: Do we need it already? It feels strange to have this and an empty Number

File: chapter-05/1_1_types/money/amount.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package money
+
+// Amount represents a given amount of money.
+type Amount struct {
+	number   Number
```
Comment (floppyzedolfin) [inline chapter-05/1_1_types/money/amount.go]: Amount needs to be exposed (returned by Convert), but why would Number and Currenecy be?

File: chapter-05/1_1_types/money/convert.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package money
+
+// exchangeRates is the dependency that fetches applicable rates.
+type exchangeRates interface {
```
Comment (floppyzedolfin) [inline chapter-05/1_1_types/money/convert.go]: Use a -er form? `exchangeRatesRetriever` ?

File: chapter-05/2_represent/money/number.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+const (
+	// ErrInvalidInteger is returned if the integer part is not a number.
+	ErrInvalidInteger = moneyError("unable to convert integer part")
+
+	// ErrInvalidDecimal is returned if the decimal part is not a number.
+	ErrInvalidDecimal = moneyError("unable to convert decimal part")
+
+	// ErrTooLarge is returned if the amount is too large - this would cause floating point precision errors.
+	ErrTooLarge = moneyError("amount over 10^15 is too large")
```
Comment (floppyzedolfin) [inline chapter-05/2_represent/money/number.go:28]: different from maxAmount.

`ErrTooLarge = moneyError(fmt.Sprintf("amount over %d is too large", maxAmount)`

File: chapter-05/2_represent/money/amount_internal_test.go [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package money
+
+import (
+	"testing"
+)
+
+func TestNewAmount(t *testing.T) {
+	// TODO !!!
```
Comment (ablqk) [inline chapter-05/2_represent/money/amount_internal_test.go:8]: Any candidate? @floppyzedolfin @doniacld

**Pull request merged**